docs/lazy-adr/adr-005-da-modules-and-components.md

```diff
@@ -0,0 +1,122 @@
+# Lazyledger ADR - Data Availability Module and Components
+
+## Changelog
+
+* 10.06.2021 - Created
+
+## Context
+The core of LazyLedger technology relies on Data Availability schema. For the moment of ADR creation, most DA pieces and
+primitives like `nmt` and `rsmt2d` are already implemented. Furthermore, prototype implementation over the network
+exists as well. However, it still has several unfilled gaps:
+* Slow data retreivability - Validators needs to get data faster and proactively.
+* Fine-grained control over dependencies and 3rd party protocols.
+* Implementation modularity - similar to `nmt` and `rsmt2d`, as it's meant to be used outside of `ll-core` as well.
+* Generic data types and abstractions - contribute to modularity and configurable composition of DA implementations.
+
+The goal of this ADR is to provide a detailed design on how to resolve issues above.
+
+## Detailed Design
+
+### Package
+Data Availability related logic is place in `lazyledger-core/p2p/ipld`. The name and location was good for original
+purpose, as that's a p2p stuff that uses IPLD in it's core. However, currently it makes sense to generalize and rename
+the package to `da`(short for DataAvailability) or even move it in it's own repository, what could be done later with
+ease if package is properly designed.
+
+### Data Types
+Currently, `DAHeader` and block's `Data` are defined in `ll-core/types`. However, with modularity in mind and plan for
+DA to be used outside of `ll-core`, e.g. in `optimint`, it now make sense to move them to new `da` package.
```

Comment:
Reiterating what I've mentioned in our call, it's non-obvious the lower-level APIs defined here need to be modular (e.g. the interfaces below). 
Optimint is a good point but the abstractions there will be much more general.

Moving the DA logic out of the types package into a separate one (maybe even into a separate repo later) is a good point though. The networking-related methods don't belong in the types package 👌🏼  

Though moving out the  `DAHeader` or the block's `Data` out of the types package will be inconsistent with the other core types but maybe there are good reasons for that.

Reply:
+1 on moving out the DA logic. Also +1 on not moving out the `Data` especially, as it's purely a detail of LazyLedger. The `DAHeader` is a bit trickier, as it straddles both domains. Could the `DAHeader` not be present in both `ll-core` and `da`?

Reply:
> Could the DAHeader not be present in both ll-core and da

It can be removed entirely in general. But for now, I would just move it to DA lib and make `ll-core` to depend on it and use DAHeader defined there.